libraries/chain/htlc_evaluator.cpp

```diff
@@ -0,0 +1,167 @@
+                esc.preimage_hash_algorithm = o.hash_type;
+             });
+             return  esc.id;
+
+          } FC_CAPTURE_AND_RETHROW( (o) )
+      }
+
+      template<typename T>
+      bool htlc_redeem_evaluator::test_hash(const std::vector<unsigned char>& incoming_preimage, const std::vector<unsigned char>& valid_hash)
+      {
+         // convert incoming_preimage to an array
+         unsigned char incoming_array[incoming_preimage.size()];
+         for(unsigned i = 0; i < incoming_preimage.size(); ++i)
+            incoming_array[i] = incoming_preimage[i];
+         T attempted_hash = T::hash( (char*)incoming_array, incoming_preimage.size());
+         if (attempted_hash.data_size() != valid_hash.size())
+            return false;
+         char* data = attempted_hash.data();
+         for(size_t i = 0; i < attempted_hash.data_size(); ++i)
+         {
+            if ( ((unsigned char)data[i]) != valid_hash[i])
+               return false;
+         }
+         return true;
+      }
+
+      void_result htlc_redeem_evaluator::do_evaluate(const htlc_redeem_operation& o)
+      {
+    	   htlc_obj = &db().get<htlc_object>(o.htlc_id);
+
+  		   FC_ASSERT(o.preimage.size() == htlc_obj->preimage_size, "Preimage size mismatch.");
+    		FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "Preimage provided after escrow expiration.");
+
+    		// see if the preimages match
+         bool match = false;
+         if (htlc_obj->preimage_hash_algorithm == graphene::chain::hash_algorithm::sha256)
+            match = test_hash<fc::sha256>(o.preimage, htlc_obj->preimage_hash);
+         if (htlc_obj->preimage_hash_algorithm == graphene::chain::hash_algorithm::ripemd160)
+            match = test_hash<fc::ripemd160>(o.preimage, htlc_obj->preimage_hash);
+         if (htlc_obj->preimage_hash_algorithm == graphene::chain::hash_algorithm::sha1)
+            match = test_hash<fc::sha1>(o.preimage, htlc_obj->preimage_hash);
+
+    		FC_ASSERT(match, "Provided preimage does not generate correct hash.");
+    	   return void_result();
+      }
+
+      void_result htlc_redeem_evaluator::do_apply(const htlc_redeem_operation& o)
+      {
+ 		   db().adjust_balance(htlc_obj->to, htlc_obj->amount);
+  		   db().remove(*htlc_obj);
+    	   return void_result();
+      }
+
+      void_result htlc_extend_evaluator::do_evaluate(const htlc_extend_operation& o)
+      {
+    	   htlc_obj = &db().get<htlc_object>(o.htlc_id);
+
+  		   FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "HTLC has already expired.");
+
+    	   return void_result();
+      }
+
+      void_result htlc_extend_evaluator::do_apply(const htlc_extend_operation& o)
+      {
+         db().modify(*htlc_obj, [&o](htlc_object& db_obj) {
+            db_obj.expiration = fc::time_point_sec(db_obj.expiration.sec_since_epoch() + o.seconds_to_add);
```

Comment:
Can add seconds directly to `fc::time_point_sec`: https://github.com/bitshares/bitshares-fc/blob/master/include/fc/time.hpp#L103